esp-wifi/src/wifi/mod.rs

```diff
@@ -1416,49 +1416,51 @@ mod asynch {
         pub async fn scan_n<const N: usize>(
             &mut self,
         ) -> Result<(heapless::Vec<AccessPointInfo, N>, usize), WifiError> {
-            critical_section::with(|cs| WIFI_EVENTS.borrow_ref_mut(cs).remove(WifiEvent::ScanDone));
-            esp_wifi_result!(crate::wifi::wifi_start_scan(false))?;
-            WifiEventFuture::new(WifiEvent::ScanDone).await;
+            let future = self.wait_for_event(WifiEvent::ScanDone);
+            esp_wifi_result!(wifi_start_scan(false))?;
+            future.await;
             self.scan_results()
         }
 
         /// Async version of [`embedded_svc::wifi::Wifi`]'s `start` method
         pub async fn start(&mut self) -> Result<(), WifiError> {
-            critical_section::with(|cs| WIFI_EVENTS.borrow_ref_mut(cs).remove(WifiEvent::StaStart));
-            wifi_start()?;
-
             let is_ap = match self.config {
                 embedded_svc::wifi::Configuration::None => panic!(),
                 embedded_svc::wifi::Configuration::Client(_) => false,
                 embedded_svc::wifi::Configuration::AccessPoint(_) => true,
                 embedded_svc::wifi::Configuration::Mixed(_, _) => panic!(),
             };
-
-            if is_ap {
-                WifiEventFuture::new(WifiEvent::ApStart).await;
+            let event = if is_ap {
+                WifiEvent::ApStart
             } else {
-                WifiEventFuture::new(WifiEvent::StaStart).await;
-            }
+                WifiEvent::StaStart
+            };
+            critical_section::with(|cs| WIFI_EVENTS.borrow_ref_mut(cs).remove(event));
+            wifi_start()?;
+
+            self.wait_for_event(event).await;
```

Comment:
`wait_for_event` will unconditionally remove the event first but the ApStart/StaStart is already set before we get here.
i.e. `start` will never complete

Changing this back to
```rust
            if is_ap {
                WifiEventFuture::new(WifiEvent::ApStart).await;
            } else {
                WifiEventFuture::new(WifiEvent::StaStart).await;
            }
```

Should make it work

Reply:
Hmm, right, I did introduce a few changes like this. Thanks for pointing it out!